d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -254,16 +334,200 @@ public void onError(Throwable e)
      * @param context Context of the retry request.
+     * @param numberOfRetryAttempts Number of retry attempts.
      * @return {@code true} if a request can be retried; {@code false} otherwise;
      */
-    public abstract boolean doRetryRequest(REQ request, RequestContext context);
+    public abstract boolean doRetryRequest(REQ request, RequestContext context, int numberOfRetryAttempts);
+  }
+
+  @ThreadSafe
+  private class ClientRetryTracker
+  {
+    private final int _aggregatedIntervalNum;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+    private final String _serviceName;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private double _currentAggregatedRetryRatio;
+
+    @GuardedBy("_counterLock")
+    private final LinkedList<RetryCounter> _retryCounter;
+    private final RetryCounter _aggregatedRetryCounter;
+
+    private ClientRetryTracker(int aggregatedIntervalNum, long updateIntervalMs, Clock clock, String serviceName)
+    {
+      _aggregatedIntervalNum = aggregatedIntervalNum;
+      _updateIntervalMs = updateIntervalMs;
+      _clock = clock;
+      _serviceName = serviceName;
+
+      _lastRollOverTime = clock.currentTimeMillis();
+      _currentAggregatedRetryRatio = 0;
+
+      _aggregatedRetryCounter = new RetryCounter();
+      _retryCounter = new LinkedList<>();
+      _retryCounter.add(new RetryCounter());
+    }
+
+    public void add(boolean isRetry)
+    {
+      synchronized (_counterLock)
+      {
+        if (isRetry)
+        {
+          _retryCounter.getLast().addToRetryRequestCount(1);
+        }
+
+        _retryCounter.getLast().addToTotalRequestCount(1);
+      }
+      updateRetryDecision();
+    }
+
+    public void rollOverStats()
+    {
+      // rollover the current interval to the aggregated counter
+      synchronized (_counterLock)
+      {
+        RetryCounter intervalToAggregate = _retryCounter.getLast();
+        _aggregatedRetryCounter.addToTotalRequestCount(intervalToAggregate.getTotalRequestCount());
+        _aggregatedRetryCounter.addToRetryRequestCount(intervalToAggregate.getRetryRequestCount());
+
+        if (_retryCounter.size() > _aggregatedIntervalNum)
+        {
+          // discard the oldest interval
+          RetryCounter intervalToDiscard = _retryCounter.removeFirst();
+          _aggregatedRetryCounter.subtractFromTotalRequestCount(intervalToDiscard.getTotalRequestCount());
+          _aggregatedRetryCounter.subtractFromRetryRequestCount(intervalToDiscard.getRetryRequestCount());
+        }
+
+        // append a new interval
+        _retryCounter.addLast(new RetryCounter());
+      }
+    }
+
+    public void isBelowRetryRatio(SuccessCallback<Boolean> callback)
+    {
+      _balancer.getLoadBalancedServiceProperties(_serviceName, new Callback<ServiceProperties>()
```

Comment:
Is this making a call out to something like ZK to get this data? If so doing this on each invocation that could add a lot of latency. Ideally the result would be cached with some TTL (or maybe just fetched once the first time - do we expect this to change in ZK?).

Reply:
It will only make a call to ZK the first time getting data from a downstream service. At the same time, it will also register a listener to ZK for property updates. The subsequent calls will just read from the local LoadBalancerState.

Reply:
Thanks for clarifying